### ISSUE TITLE
Support flatpak installations of heroic (e.g. on Steam Deck)

### ISSUE THREAD

**Issue by sluedecke** [mtkennerly/ludusavi#142]
Regarding Steam Deck: though it works for me, it does not work well for a installation of heroic done via flatpak.

#### A) Installation with standard Linux tools

I installed heroic manually as a normal package with pacman on the deck.  This involves disabling read only for the Steam Deck file system and is something the majority of users will not do.  Still it creates a situation on the deck which is similar to a normal installation on any other machine and heroic configuration is found in `~/.config/heroic`.
In this situation ludusavi finds the save games in all the known / expected locations.

#### B) Flatpak installation

However, if one chooses to install heroic via flatpak, heroic configuration ends up in `~/.var/app/com.heroicgameslauncher.hgl/config/heroic`.  The problem is with some Linux native games, e.g. Terraria and Spiritfarer.  They then store their save games in `.var/app/com.heroicgameslauncher.hgl/data/Terraria`  (instead of `~/.local/share/Terraria`) and `.var/app/com.heroicgameslauncher.hgl/config/unity3d/Thunder Lotus Games/Spiritfarer`. And ludusavi does not find these.  Windows games seem to work fine so far.

So for a flatpak installation more work needs to be done.

@mtkennerly if you agree, I would work to finish this item for situation A and create a new issue to extend it to properly handle B once A is done.

_Originally posted by @sluedecke in https://github.com/mtkennerly/ludusavi/issues/94#issuecomment-1284998151_

**Comment by LeSnake04** [comment]
The thing is: Most games use /home/user even when installed trough heroic / steam Flatpak. Only some actually use the flatpak sandbox of the store.

IMO the way to go for B is a Root called "flatpak home" thats under the box is handled like a Home, but with .config being replaced with config and .local/share replaced by data.

Then find roots should add
var/com.valvesoftware.steam
var/com.heroicgameslauncher.hgl

Or, which might be easier: Add the flatpak home root and add a root named "Group of flatpak roots", which adds all subfolders (maybe with a basic validation of the structure: does the subfolder contain directories named config and data) as "flatpak root"

**Comment by mtkennerly** [comment]
> The thing is: Most games use /home/user even when installed trough heroic / steam Flatpak. Only some actually use the flatpak sandbox of the store.

The `prelude::parse_paths` change I suggested would just be an additional check. It would still look for the normal locations as well.

I'd rather try to handle it transparently like that, rather than requiring an extra Flatpak root type, if we can avoid it.

**Comment by LeSnake04** [comment]
(just added in an edit)
> That way we also can find saves for native linux games installed via flatpak.

That would a nice side-effect from solving it this way IMO

**Comment by LeSnake04** [comment]
And saves from other launchers might be detected as well.

**Comment by mtkennerly** [comment]
> That way we also can find saves for native linux games installed via flatpak.

Oh, that's an interesting idea. Do you have any specific examples of these offhand? I'm curious if the config/data folder naming is consistent and how the Flatpak install folders themselves compare to the game titles.

We might be able to start checking these automatically without any extra config, although it would probably be prohibitively slow unless we can identify the specific Flatpak install for the specific game. We might be able to do some fuzzy folder matching like we do for normal install folders.

> (different topic: Something like "group of wineprefixes" might be useful as well)

You can do this already with globs :) Just make it a Wine prefix root and set the path to `/prefix/group/*`. It will make the scan a lot slower, though, because it has to check every prefix for every game.

**Comment by LeSnake04** [comment]
https://flathub.org/apps/details/org.supertuxproject.SuperTux

Its looks just like when installed via pacman/apt etc. Its just located in .var/app/org.supertuxproject.SuperTux/data/supertux2 instead of .local/share/supertux

> lsd .var/app/org.supertuxproject.SuperTux/data/ --tree                                                                                                                               11:21:57
 data
└──  supertux2
    ├──  addons
    │   ├──  Arons_Levels.zip
    │   ├──  ATTNW1.zip
    │   ├──  ATTNW2.zip
    │   ├──  cschaefer-tux-strikes-back.zip
    │   ├──  ctdabomb-worldmaps.zip
    │   ├──  grandmaster-journey.zip
    │   ├──  hume-worldmaps.zip
    │   ├──  language-pack.zip
    │   ├──  lmh-fionas-world.zip
    │   ├──  lmh-matties-world.zip
    │   ├──  lolkat-levels.zip
    │   ├──  manuel_desert_world.zip
    │   ├──  michel-the-frozen-mountains.zip
    │   ├──  narre-land.zip
    │   ├──  niso-yetis-revenge.zip
    │   ├──  octo-levels.zip
    │   ├──  repository.nfo
    │   ├──  rubberduck-levels.zip
    │   ├──  semajd-valley-of-chaos.zip
    │   ├──  serano-islands.zip
    │   ├──  The_Big_Island.zip
    │   ├──  toufik-levels.zip
    │   ├──  Treasure_Mountain.zip
    │   ├──  treskalle-island-of-moob.zip
    │   └──  wolf-levels.zip
    ├──  config
    ├──  levels
    │   ├──  bonus4
    │   │   └──  A_Narrow_Path.stl
    │   ├──  Easy_Mode
    │   │   ├──  info
    │   │   └──  level1.stl
    │   └──  manuel_desert_world
    │       └──  bonuslevel.stl
    └──  profile1
        ├──  arons_levels.stsg
        ├──  ATTNW1.stsg
        ├──  ATTNW2.stsg
        ├──  bonus1.stsg
        ├──  bonus2.stsg
        ├──  bonus3.stsg
        ├──  bonus4.stsg
        ├──  christmas2016.stsg
        ├──  community2016.stsg
        ├──  CT's Domain.stsg
        ├──  Easy_Mode.stsg
        ├──  fiona.stsg
        ├──  grandmaster-journey.stsg
        ├──  halloween2014.stsg
        ├──  hume.stsg
        ├──  hume_rainbow_cave.stsg
        ├──  levels-of-wolfs.stsg
        ├──  Lolkat's Levels.stsg
        ├──  manuel_desert_world.stsg
        ├──  mattie_world.stsg
        ├──  monde.stsg
        ├──  moob-island.stsg
        ├──  narre.stsg
        ├──  penny.stsg
        ├──  revenge_in_redmond.stsg
        ├──  seranos-islands.stsg
        ├──  The Big Island.stsg
        ├──  The_Valley_Of_Chaos.stsg
        ├──  Toufiks levels.stsg
        ├──  Treasure_Mountain.stsg
        ├──  tsb.stsg
        ├──  world1.stsg
        └──  yetis_revenge.stsg

**Comment by LeSnake04** [comment]
I looked into it. PcGamingwiki doesnt support flathub as store (Making a feature request at the moment). Until then the only way is hardcodind, searching all of them or making a custom list

**Comment by LeSnake04** [comment]
https://community.pcgamingwiki.com/topic/5020-request-add-flathub-as-source/

**Comment by LeSnake04** [comment]
Or we only search flatpaks tagged as games on flathub: https://flathub.org/api/v1/apps/category/Games

**Comment by mtkennerly** [comment]
That would be great if the Flatpak IDs were added to PCGW. I don't know how active the on-wiki discussions are, but you might also want to consider posting on their Discord, which is pretty active.

**Comment by mtkennerly** [comment]
I created a separate ticket for the non-Heroic Flatpak case: https://github.com/mtkennerly/ludusavi/issues/143

**Comment by LeSnake04** [comment]
Maybe it makes sense to implement the system for getting flatpaks defined by the manifest and just manually define the heroic and steam id for now.

**Comment by LeSnake04** [comment]
I would code it like this:

+ for id in ids_in_manifest
  + if <flatpak_root(default: .var)>/app/<id>
    + add home with replaced paths

**Comment by mtkennerly** [comment]
@LeSnake04 I responded in #143.

**Comment by mtkennerly** [comment]
@sluedecke Would you be able to take a look at the Heroic Flatpak support? I added Windows support, so this should be the last thing we need.

**Comment by sluedecke** [comment]
Yes, I will check if I can add some tests for #94, then continue with this one.

**Comment by sluedecke** [comment]
So far the suggested `prelude::parse_paths` seems to do the job, but there is still some work left to do to handle all situations properly and to avoid unnecessary scans.

**Comment by mtkennerly** [comment]
Done in #149 🎉
